File: clients/ios/demo/Sources/ClientToolsDemo/Network/AuthService.swift
import Foundation
import ClientToolsSDK

class AuthService {
    static let baseURL = "http://api.pulse.app/"

    private let session: URLSession

    init() {
        session = ClientToolsSDK.shared.makeMockSession()
    }

    func sendSmsCode(phone: String) async throws -> BaseResponse {
        try await post("auth/sms/send", body: SendSmsRequest(phone: phone))
    }

    func loginSms(phone: String, code: String) async throws -> LoginResponse {
        try await post("auth/login/sms", body: SmsLoginRequest(phone: phone, code: code))
    }

    func loginPassword(phone: String, password: String) async throws -> LoginResponse {
        try await post("auth/login/password", body: PasswordLoginRequest(phone: phone, password: password))
    }

    func loginEmail(email: String, password: String) async throws -> LoginResponse {
        try await post("auth/login/email", body: EmailLoginRequest(email: email, password: password))
    }

    private func post<Req: Encodable, Resp: Decodable>(_ path: String, body: Req) async throws -> Resp {
        var request = URLRequest(url: URL(string: AuthService.baseURL + path)!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(Resp.self, from: data)
    }
}

File: clients/android/demo/src/main/kotlin/com/clienttools/demo/network/AuthService.kt
package com.clienttools.demo.network

import com.clienttools.demo.model.BaseResponse
import com.clienttools.demo.model.EmailLoginRequest
import com.clienttools.demo.model.LoginResponse
import com.clienttools.demo.model.PasswordLoginRequest
import com.clienttools.demo.model.SendSmsRequest
import com.clienttools.demo.model.SmsLoginRequest
import retrofit2.http.Body
import retrofit2.http.POST

interface AuthService {
    @POST("auth/sms/send")
    suspend fun sendSmsCode(@Body req: SendSmsRequest): BaseResponse

    @POST("auth/login/sms")
    suspend fun loginSms(@Body req: SmsLoginRequest): LoginResponse

    @POST("auth/login/password")
    suspend fun loginPassword(@Body req: PasswordLoginRequest): LoginResponse

    @POST("auth/login/email")
    suspend fun loginEmail(@Body req: EmailLoginRequest): LoginResponse
}
